llvm-external-projects/iree-dialects/lib/Dialect/LinalgExt/IR/LinalgExtOps.cpp

```diff
@@ -1447,6 +1484,388 @@ LogicalResult TopkOp::getResultTilePosition(
+            ceilDiv(inputType.getDimSize(i), tileAndPosMapping[i]);
+        inferredShape.push_back(sizeDim);
+      }
+    } else
+      inferredShape.push_back(inputType.getShape()[i]);
+  }
+
+  // point loop.
+  SmallVector staticTiles = getStaticTiles();
+  inferredShape.append(staticTiles.begin(), staticTiles.end());
+
+  return TypeSwitch<Type, ShapedType>(inputType)
+      .Case<RankedTensorType>([&](RankedTensorType t) -> ShapedType {
+        return RankedTensorType::get(inferredShape, inputType.getElementType());
+      })
+      .Case<MemRefType>([&](MemRefType t) -> ShapedType {
+        return MemRefType::get(inferredShape, inputType.getElementType());
+      })
+      .Default([&](Type t) {
+        llvm_unreachable("unexpected type");
+        return nullptr;
+      });
+}
+
+// Return true if at least one element in `tiles` is zero.
+static bool hasZeros(ArrayRef<OpFoldResult> tiles) {
+
+  // Check if `tile` is an integer zero attribute or
+  // comes from a zero constantIndexOp.
+  auto isZero = [](OpFoldResult tile) -> bool {
+    if (Attribute attr = tile.dyn_cast<Attribute>()) {
+      IntegerAttr intAttr = attr.dyn_cast<IntegerAttr>();
+      return intAttr && intAttr.getValue().isZero();
+    } else if (auto cst =
+                   tile.get<Value>().getDefiningOp<arith::ConstantIndexOp>())
+      return cst.value() == 0;
+    return false;
+  };
+
+  return llvm::any_of(tiles, [&](OpFoldResult tile) { return isZero(tile); });
+}
+
+// Return true if `dimsPos` is invalid. It is invalid when: a) it contains
+// duplicate.
+static bool isInvalid(ArrayRef<int64_t> dimsPos) {
+  DenseSet<int64_t> uniqued;
+  for (int64_t dim : dimsPos)
+    uniqued.insert(dim);
+  return dimsPos.size() != uniqued.size();
+}
+
+// Check if we have enough static information to catch undefined behavior when
+// the tile size does not divide perfectly the dimension of the input tensor.
+static bool areNotFullTiles(ArrayRef<int64_t> inputShape,
+                            DenseMap<int64_t, int64_t> dimAndTileMapping) {
+  int64_t rank = inputShape.size();
+  for (int64_t dim = 0; dim < rank; dim++) {
+    // dynamic dimension.
+    if (inputShape[dim] == ShapedType::kDynamicSize)
+      continue;
+    // static dimension but dynamic tile factor.
+    if (dimAndTileMapping.count(dim) &&
+        dimAndTileMapping[dim] == ShapedType::kDynamicSize)
+      continue;
+    // static tile and dimension, check.
+    if (dimAndTileMapping.count(dim) &&
+        inputShape[dim] % dimAndTileMapping[dim] != 0)
+      return true;
+  }
+  return false;
+}
+
+// verifier for the pack operation.
+LogicalResult PackOp::verify() {
+  Operation *op = getOperation();
+  size_t numberOfBlockingFactors = getMixedTiles().size();
+  SmallVector<int64_t> dimsPos = extractFromI64ArrayAttr(getDimsPos());
+  // Blocking factors must be less or equal than the input rank, and must
+  // match the number of `dims_pos`.
+  if (numberOfBlockingFactors > getInputRank())
```

Comment:
Nit : if body is multiple lines use `{` and `}` (that is LLVM style AFAIK). Here and everywhere else.